core/provider/digitalocean/provider.go

```diff
@@ -0,0 +1,365 @@
+	if err := definition.ValidateBasic(); err != nil {
+		return nil, fmt.Errorf("failed to validate task definition: %w", err)
+	}
+
+	if definition.ProviderSpecificConfig == nil {
+		return nil, fmt.Errorf("digitalocean specific config is nil for %s", definition.Name)
+	}
+
+	var doConfig DigitalOceanTaskConfig
+	doConfig, ok := definition.ProviderSpecificConfig.(DigitalOceanTaskConfig)
+	if !ok {
+		return nil, fmt.Errorf("invalid provider specific config type for %s", definition.Name)
+	}
+
+	if err := doConfig.ValidateBasic(); err != nil {
+		return nil, fmt.Errorf("could not cast digitalocean specific config: %w", err)
+	}
+
+	p.logger.Info("creating droplet", zap.String("name", definition.Name))
+
+	droplet, err := p.CreateDroplet(ctx, definition)
+	if err != nil {
+		return nil, err
+	}
+
+	ip, err := droplet.PublicIPv4()
+	if err != nil {
+		return nil, err
+	}
+
+	p.logger.Info("droplet created", zap.String("name", droplet.Name), zap.String("ip", ip))
+
+	if p.dockerClients[ip] == nil {
+		p.dockerClients[ip], err = NewDockerClient(fmt.Sprintf("tcp://%s:%s", ip, sshPort))
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	_, _, err = p.dockerClients[ip].ImageInspectWithRaw(ctx, definition.Image.Image)
+	if err != nil {
+		p.logger.Info("image not found, pulling", zap.String("image", definition.Image.Image))
+		err = pullImage(ctx, p.dockerClients[ip], p.logger, definition.Image.Image)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	_, err = p.dockerClients[ip].ContainerCreate(ctx, &container.Config{
+		Image:      definition.Image.Image,
+		Entrypoint: definition.Entrypoint,
+		Cmd:        definition.Command,
+		Tty:        false,
+		Hostname:   definition.Name,
+		Labels: map[string]string{
+			providerLabelName: p.state.Name,
+		},
+		Env: convertEnvMapToList(definition.Environment),
+	}, &container.HostConfig{
+		Mounts: []mount.Mount{
+			{
+				Type:   mount.TypeBind,
+				Source: "/docker_volumes",
+				Target: definition.DataDir,
+			},
+		},
+		NetworkMode: container.NetworkMode("host"),
+	}, nil, nil, definition.ContainerName)
+	if err != nil {
+		return nil, err
+	}
+
+	taskState := &TaskState{
+		ID:           droplet.ID,
+		Name:         definition.Name,
+		Definition:   definition,
+		Status:       provider.TASK_STOPPED,
+		ProviderName: p.state.Name,
+	}
+
+	p.stateMu.Lock()
+	defer p.stateMu.Unlock()
+
+	p.state.TaskStates[taskState.ID] = taskState
+
+	return &Task{
+		state:        taskState,
+		provider:     p,
+		sshKeyPair:   p.state.sshKeyPair,
+		logger:       p.logger.With(zap.String("task", definition.Name)),
+		doClient:     p.doClient,
+		dockerClient: p.dockerClients[ip],
+	}, nil
+}
+
+func (p *Provider) SerializeProvider(context.Context) ([]byte, error) {
+	p.stateMu.Lock()
+	defer p.stateMu.Unlock()
+
+	bz, err := json.Marshal(p.state)
+
+	return bz, err
+}
+
+func (p *Provider) DeserializeProvider(context.Context) ([]byte, error) {
```

Comment:
what's the point of this function - is this the equivalent of RestoreProvider? it's the same as SerializeProvider